### ISSUE TITLE
Removal of semicolon in notebooks

### ISSUE THREAD

**Issue by peendebak** [astral-sh/ruff#7300]
One of the default rules in ruff is to remove a trailing semicolon. That is fine for python scripts, for for jupyter notebook this can change the output of the code cells. A screenshot to illustrate:

![image](https://github.com/astral-sh/ruff/assets/17724047/8ee72dcd-de66-418c-83d5-fc43e98e0b01)

The removal of the semicolon results in the output of the `plt.plot` command being displayed. Typically this is not the intention, so a semicolor is used to hide the command output and only the the plotted graph.

Can the ruff default settings be modified to remove semicolons for python files, but not for notebooks?

**Comment by dhruvmanila** [comment]
I think this might be similar to #6586 where it's only the last expression in a cell where this rule should be disabled. We could probably have a general interface internally to include rules which need not be checked for last expressions in all cells within a Jupyter Notebook.

Until that is supported, you could disable this rule on Jupyter Notebooks like so:
```toml
[tool.ruff]
per-file-ignores = { "*.ipynb" = ["E703"] }
```

https://beta.ruff.rs/docs/rules/useless-semicolon/

**Comment by henryiii** [comment]
There are several rules that don't apply to notebooks, or apply in a modified form. `B018` - A useless expression at the end of a cell is actually not useless, that's how you show the (notebook customized if applicable) display of an object. `T20` also usually doesn't apply, notebooks are not library files, but I think it's fine to have to disable that.

Ideally `B018` and `E703` would learn that when in a cell, they don't apply to the last item in the cell.

**Comment by charliermarsh** [comment]
Agreed.

**Comment by charliermarsh** [comment]
We already had to solve this cell-boundaries problem for isort, so should be doable in these cases too.

**Comment by s-weigand** [comment]
> ```toml
> [tool.ruff]
> per-file-ignores = { "*.ipynb" = ["E703"] }
> ```
Sadly this seems not to be respected by `ruff format` 😢

**Comment by dhruvmanila** [comment]
(Merging this with a general tracking issue: https://github.com/astral-sh/ruff/issues/8669)

**Comment by dhruvmanila** [comment]
> > ```toml
> > [tool.ruff]
> > per-file-ignores = { "*.ipynb" = ["E703"] }
> > ```
> 
> Sadly this seems not to be respected by `ruff format` 😢 Even so, it is not part of [the conflicting rules](https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules)

@s-weigand Hi, we recently updated the formatter to not remove any trailing semicolon in notebook context: https://github.com/astral-sh/ruff/pull/8590. It'll be available in the next release.

**Comment by Norlandz** [comment]
https://github.com/astral-sh/ruff/issues/7300#issuecomment-1809539473

> > > [tool.ruff]
> > > per-file-ignores = { "*.ipynb" = ["E703"] }
> > 
> > 
> > Sadly this seems not to be respected by `ruff format` 😢 Even so, it is not part of [the conflicting rules](https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules)
> 
> [@s-weigand](https://github.com/s-weigand) Hi, we recently updated the formatter to not remove any trailing semicolon in notebook context: [#8590](https://github.com/astral-sh/ruff/pull/8590). It'll be available in the next release.

It seems **only the last line is respected** with that? 

Before
```py
# Plotting
x = torch.arange(len(vocab))
bar_width = 0.15

fig, ax = plt.subplots(figsize=(5, 3))
for i, T in enumerate(temperatures):
    rects = ax.bar(x + i * bar_width, scaled_probas[i], bar_width, label=f'Temperature = {T}')

ax.set_ylabel('Probability');
ax.set_xticks(x);
ax.set_xticklabels(vocab.keys(), rotation=90);
ax.legend();

plt.tight_layout();
plt.savefig("temperature-plot.pdf");
plt.show();
```

After
```py
# Plotting
x = torch.arange(len(vocab))
bar_width = 0.15

fig, ax = plt.subplots(figsize=(5, 3))
for i, T in enumerate(temperatures):
    rects = ax.bar(x + i * bar_width, scaled_probas[i], bar_width, label=f"Temperature = {T}")

ax.set_ylabel("Probability")
ax.set_xticks(x)
ax.set_xticklabels(vocab.keys(), rotation=90)
ax.legend()
plt.tight_layout()
plt.savefig("temperature-plot.pdf")
plt.show();
```

Is there a way to just tell ruff, when inside ipynb, just **not touch semicolon at all**? And dont trim my newline for that?

The reason I use **multiple semicolon** is that, I use `get_ipython().ast_node_interactivity = "all"` to print multiple outputs in a cell. 

Related:

- https://github.com/astral-sh/ruff/issues/8254

- https://github.com/astral-sh/ruff/pull/8590

**Comment by dhruvmanila** [comment]
> It seems **only the last line is respected** with that?

Yes, that was intentional. The main reason was that a semicolon on the last line is usually intentional to avoid having the cell output corresponding to that last line.

> The reason I use **multiple semicolon** is that, I use `get_ipython().ast_node_interactivity = "all"` to print multiple outputs in a cell.

Thanks for this information. Looking at my notes when implementing this change, I think I knew about this option but the change was specific to accommodate the `last_expr_or_assign` value for that option for which I'm not sure what the motivation was.

> Is there a way to just tell ruff, when inside ipynb, just **not touch semicolon at all**? And dont trim my newline for that?

I think the first (not touch semicolon at all) is a reasonable request but I'm not sure what do you mean by "don't trim my newline for that"? Do you mind opening a new issue for the first request?

**Comment by Norlandz** [comment]
@dhruvmanila Thanks for the reply.

> > Is there a way to just tell ruff, when inside ipynb, just **not touch semicolon at all**? And dont trim my newline for that?
> 
> I think the first (not touch semicolon at all) is a reasonable request but I'm not sure what do you mean by "don't trim my newline for that"? Do you mind opening a new issue for that?

For "don't trim my newline".
I mean this:

Before
```
ax.set_xticklabels(vocab.keys(), rotation=90);
ax.legend();

plt.tight_layout();
plt.savefig("temperature-plot.pdf");
```

After
```
ax.set_xticklabels(vocab.keys(), rotation=90)
ax.legend()
plt.tight_layout()
plt.savefig("temperature-plot.pdf")
```

That is not a big issue to me, but just saying.

**Comment by Norlandz** [comment]
@dhruvmanila 

Actually **Dont put too much time on this**. 

I dont think semicolons works well with `get_ipython().ast_node_interactivity = "all"`
I was using `_ =` all the time to suppress the output. 
I though `;` works the same but no...

For example:

![Image](https://github.com/user-attachments/assets/1133aad2-972d-4ebe-897f-ed0412308d35)

For the following, the `bb;` clearly didnt get suppressed, so semicolon was not helpful.
![Image](https://github.com/user-attachments/assets/f718f4f0-58bf-4e0e-b284-e0a5b713a5f0)

ipynb file for testing:
```
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.getipython import get_ipython\n",
    "\n",
    "aa = \"hello\"\n",
    "bb = \"foo\"\n",
    "cc = \"bar\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_ipython().ast_node_interactivity = \"last_expr\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'bar'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa\n",
    "bb\n",
    "cc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'bar'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa\n",
    "bb;\n",
    "cc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa\n",
    "bb\n",
    "cc;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_ipython().ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'hello'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "'foo'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "'bar'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa\n",
    "bb\n",
    "cc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'hello'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "'foo'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "'bar'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa\n",
    "bb;\n",
    "cc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa\n",
    "bb\n",
    "cc;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa\n",
    "_ = bb\n",
    "cc"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
```

**Comment by dhruvmanila** [comment]
Sounds good, thanks for testing and getting back :)
